### PR TITLE
Simplify Python FunctionSpace creation

### PR DESCRIPTION
**Pull request by garth-wells**
- Simplifies FunctionSpace initialiser
- Remove inelegant re-construction of UFL FiniteElements from string signatures from cpp objects
- Interface change to require 'bundling' of element family string and degree into into a list/tuple (leads to simplification of initialiser)

Also adds `legacy' module (name can be changed) to support old syntax.

### CODE REVIEW

File: python/dolfin/function/functionspace.py [hunk 1]
```diff
@@ -5,43 +5,50 @@
...
 
 
+class ElementDef(typing.NamedTuple):
+    """Data for representing a finite element"""
+    family: str
+    degree: int
+    form_degree: typing.Optional[int] = None  # noqa
+
+
 class FunctionSpace(ufl.FunctionSpace):
```
Comment (michalhabera) [inline python/dolfin/function/functionspace.py]: I am not sure why `dolfin.function.FunctionSpace` should inherit from `ufl.FunctionSpace`. 
Its not directly related to this issue, but better think about it. We should probably eliminate inheritance wherever possible.

Comment (garth-wells) [reply python/dolfin/function/functionspace.py]: Agree it should be considered (as a separate issue).

File: python/dolfin/function/functionspace.py [hunk 2]
```diff
@@ -5,43 +5,50 @@
...
-        # FIXME: This includes some hacks to support construction of
-        # function spaces with existing cpp FunctionSpace, which is
-        # typical when extracting subspaces.
-        #
-        # It would element if an element argument was a single object,
-        # e.g. a ufl.FiniteElement or a tuple(family degree)
-
+        # Create function space from a UFL element and existing cpp
+        # FunctionSpace
         if cppV:
```
Comment (jhale) [inline python/dolfin/function/functionspace.py]: `if cppV is not None:` ?

File: python/dolfin/function/functionspace.py [hunk 3]
```diff
@@ -5,43 +5,50 @@
...
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 
+import typing
+
 import ufl
 from dolfin import cpp, jit
 from dolfin.fem import dofmap
 
 
+class ElementDef(typing.NamedTuple):
```
Comment (jhale) [inline python/dolfin/function/functionspace.py]: `ElementDefinition`?

File: python/dolfin/function/functionspace.py [hunk 4]
```diff
@@ -5,43 +5,50 @@
...
+    """Data for representing a finite element"""
+    family: str
+    degree: int
+    form_degree: typing.Optional[int] = None  # noqa
```
Comment (jhale) [inline python/dolfin/function/functionspace.py]: Doesn't seem complete without a cell type? I the cell_type was `None`, FunctionSpace could infer it from the mesh?

Comment (garth-wells) [reply python/dolfin/function/functionspace.py]: It's the necessary meta-data.  If the element is fully defined, one case just pass an element.

**Pull request merged**